redash/utils/__init__.py

```diff
@@ -48,19 +49,28 @@ def slugify(s):
     return re.sub("[^a-z0-9_\-]+", "-", s.lower())
 
 
-def gen_query_hash(sql):
+def gen_query_hash(sql, data_source=None, set_auto_limit=False):
     """Return hash of the given query after stripping all comments, line breaks
     and multiple spaces, and lower casing all text.
 
     TODO: possible issue - the following queries will get the same id:
         1. SELECT 1 FROM table WHERE column='Value';
         2. SELECT 1 FROM table where column='value';
     """
+    if data_source is not None:
+        sql = data_source.query_runner.apply_auto_limit(sql, set_auto_limit)
```

Comment:
This is heading in a good direction! I still feel like this abstraction is leaking though.

`gen_query_hash` was previously a happy contextless function, floating in a contextless space. But now, it depends on more context and its implementation varies in different query runners.

Why not move `gen_query_hash` to `QueryRunnerBase`, with the previous implementation as default, and a polymorphic override where auto limits apply? (i.e `super.gen_query_hash(sql_after_applying_limit)`)

This will help reducing the sprawl of booleans moving between functions. WDYT?

Reply:
I agree that this logic shouldn't be here: we should resolve the correct SQL to hash before calling this function.

As for calculating the hash in the Query Runner: we need it beforehand because we use it for the lock. Going back to my note about a bigger refactor: we should probably lock on query id vs. query text. The query text locking was an optimization that these days I doubt is needed.

Reply:
> we need it beforehand because we use it for the lock

If you mean [this](https://github.com/getredash/redash/blob/default_limit/redash/tasks/queries/execution.py#L33), then the data source is available there and we can ask for it to calculate the hash, no?

Reply:
quick question: Will every query always have a data source? If yes, then I think we can move this function to query_runner. And I agree with Omer that the "we need it beforehand because we use it for the lock" won't be a problem because we are just going to call this function from query_runner instead of utils

Reply:
Just updated the code and put this logic into the query runner. 
Note: In several places I am using `BaseQueryRunner({}).gen_query_hash(querytext, False)` instead of using the query's own datasource queryRunner because in that kinds of places, we have already applied the limit, so to keep logic simple, I just use gen_query_hash from BaseQueryRunner. Otherwise, we have to do something like 
```
if self.data_source is not None:
        query_runner = self.data_source.query_runner
else:
        query_runner = BaseQueryRunner({})
self.query_hash = query_runner.gen_query_hash(self.query_text, False)
```
Since we already know that the second argument will always be False, there is no need to do this check.